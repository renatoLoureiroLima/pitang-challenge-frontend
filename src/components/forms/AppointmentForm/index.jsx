import { Formik, Form, Field, ErrorMessage } from 'formik';
import { toast } from 'react-toastify';
import { setMinutes, setHours } from 'date-fns';
import DateInput from '../../input/DateInput';
import { schedulingFormSchema } from '../../../services/validations/forms';
import axios from '../../../services/api/index';

const initialValues = {
  name: '',
  birthday: null,
  selectedDate: null,
};

const filterPassedTime = (time) => {
  const currentDate = new Date();
  const selectedDate = new Date(time);

  return currentDate.getTime() < selectedDate.getTime();
};

const onSubmit = async (values) => {
  await axios
    .post('/appointment', {
      name: values.name,
      birthday: values.birthday,
      selectedDate: values.selectedDate,
    })
    .then(() => {
      toast.info('Agendamento concluído!');
    })
    .catch((e) => {
      toast.error(e.response.data.message);
    });
};

const now = new Date();

const AppointmentForm = () => (
  <Formik
    initialValues={initialValues}
    validationSchema={schedulingFormSchema}
    onSubmit={onSubmit}
  >
    {(formik) => (
      <Form className='form'>
        <div className='form-group form-focus'>
          <label className='focus-label' htmlFor='name'>
            Nome
          </label>
          <Field
            className={`form-control floating ${formik.errors.name && 'error'}`}
            type='text'
            id='name'
            name='name'
            data-testid='nameform-test'
          />
          <ErrorMessage
            component='span'
            className='error'
            name='name'
            data-testid='error-name-test'
          />
        </div>

        <DateInput
          name='birthday'
          label='Data de Nascimento '
          maxDate={new Date()}
          data-testid='birthdayform-test'
        />

        <DateInput
          name='selectedDate'
          label='Data para Vacinação '
          showTimeSelect
          minDate={new Date()}
          minTime={setHours(setMinutes(now, 0), 6)}
          maxTime={setHours(setMinutes(now, 0), 18)}
          timeIntervals={60}
          filterTime={filterPassedTime}
          data-testid='dateform-test'
        />

        <div className='d-flex flex-row-reverse'>
          <button
            className='btn btn-primary btn-md'
            type='submit'
            disabled={!formik.dirty && formik.isValid}
            data-testid='buttonform-test'
          >
            Agendar
          </button>
        </div>
      </Form>
    )}
  </Formik>
);

export default AppointmentForm;
