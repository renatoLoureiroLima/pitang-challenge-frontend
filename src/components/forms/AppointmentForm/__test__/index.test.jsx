import ReactDOM from 'react-dom';
import AppointmentForm from '../index';

import {
  render,
  act,
  fireEvent,
  screen,
  cleanup,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import api from '../../../../services/api';

jest.mock('../../../../services/api');

afterEach(cleanup);

describe('Appointment Form', () => {
  describe('with valid inputs', () => {
    test('calls the onSubmit function', async () => {
      const mockOnSubmit = api.post.mockResolvedValue();
      await render(<AppointmentForm />);

      await act(async () => {
        fireEvent.change(screen.getByLabelText('Nome'), {
          target: { value: 'Ronaldo Vanderlei' },
        });
        fireEvent.change(screen.getByLabelText('Data de Nascimento'), {
          target: {
            value: new Date(
              'Sun Jun 13 1999 00:00:00 GMT-0300 (Horário Padrão de Brasília)'
            ),
          },
        });
        fireEvent.change(screen.getByLabelText('Data para Vacinação'), {
          target: {
            value: new Date(
              'Fri May 28 2021 15:00:00 GMT-0300 (Horário Padrão de Brasília)'
            ),
          },
        });
      });

      await act(async () => {
        userEvent.click(screen.getByText('Agendar'));
      });

      expect(mockOnSubmit).toHaveBeenCalled();
    });
  });
});

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<AppointmentForm />, div);
});

it('matches snapshot', () => {
  const tree = renderer.create(<AppointmentForm />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('invalid name', async () => {
  const { getByLabelText } = render(<AppointmentForm />);

  await act(async () => {
    const nameInput = getByLabelText('Nome');
    fireEvent.change(nameInput, { target: { value: '@7Lucas' } });
    fireEvent.blur(nameInput);
  });
  expect(screen.getByTestId('nameform-test')).toHaveValue('@7Lucas');
  expect(screen.queryByTestId('error-name-test')).toBeInTheDocument();
});
