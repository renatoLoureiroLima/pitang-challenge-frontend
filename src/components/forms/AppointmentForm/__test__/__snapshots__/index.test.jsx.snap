// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<form
  action="#"
  className="form"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="form-group form-focus"
  >
    <label
      className="focus-label"
      htmlFor="name"
    >
      Nome
    </label>
    <input
      className="form-control floating undefined"
      data-testid="nameform-test"
      id="name"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group form-focus"
  >
    <label
      className="focus-label"
      htmlFor="birthday"
    >
      Data de Nascimento 
    </label>
    <div
      className="react-datepicker-wrapper"
    >
      <div
        className="react-datepicker__input-container"
      >
        <input
          autoComplete="off"
          className="form-control floating undefined"
          disabled={false}
          id="birthday"
          name="birthday"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={false}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="form-group form-focus"
  >
    <label
      className="focus-label"
      htmlFor="selectedDate"
    >
      Data para Vacinação 
    </label>
    <div
      className="react-datepicker-wrapper"
    >
      <div
        className="react-datepicker__input-container"
      >
        <input
          autoComplete="off"
          className="form-control floating undefined"
          disabled={false}
          id="selectedDate"
          name="selectedDate"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={false}
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="d-flex flex-row-reverse"
  >
    <button
      className="btn btn-primary btn-md"
      data-testid="buttonform-test"
      disabled={true}
      type="submit"
    >
      Agendar
    </button>
  </div>
</form>
`;
